python/prophet/numpyro_model.py

```diff
@@ -0,0 +1,138 @@
+from jax import jit
+from jax.lax import scan
+import jax.numpy as jnp
+from jax.scipy.special import expit
+import numpy as np
+import numpyro.distributions as dist
+from numpyro.infer import MCMC, NUTS
+from numpyro import sample, plate, deterministic
+from numpyro.infer import SVI
+from numpyro.infer.autoguide import AutoDelta, Trace_ELBO
+from numpyro.infer.initialization import init_to_value
+from numpyro.optim import Minimize
+from jax import random
+
+
+def construct_changepoint_matrix(t, t_change, T, S):
+    A = np.zeros((T, S))
+    a_row = np.zeros(S)
+    cp_idx = 1 if S > 1 else 0
+    for i in range(T):
+        while (cp_idx < S) and (t[i] >= t_change[cp_idx]):
+            a_row[cp_idx] = 1
+            cp_idx += 1
+        A[i] = a_row
+    return jnp.array(A)
+
+
+@jit
+def linear_trend(k: float, m: float, delta: jnp.ndarray, t: jnp.ndarray, A: jnp.ndarray, t_change: jnp.ndarray):
+    first_term = jnp.multiply((k + A.dot(delta)), t)
+    second_term = (m + A.dot(jnp.multiply(-t_change, delta)))
+    return first_term + second_term
+
+
+def transition_function(carry, index):
+    m_pr, t_change, k_s = carry
+    gamma_i = (t_change[index] - m_pr) * (1 - k_s[index] / k_s[index + 1])
+    return (m_pr + gamma_i, t_change, k_s), gamma_i
+
+
+def logistic_gamma(k, m, delta, t_change, S):
+    k_s = jnp.append(jnp.array(k), k + jnp.cumsum(delta))
+    _, gamma = scan(transition_function, (m, t_change, k_s), jnp.arange(S))
+    return gamma
+
+
+def logistic_trend(k, m, delta, t, cap, A, t_change, S):
+    gamma = logistic_gamma(k, m, delta, t_change, S)
+    inv_logit = expit(jnp.multiply(k + A.dot(delta), (t - (m + A.dot(gamma)))))
+    return jnp.multiply(cap, inv_logit)
+
+
+@jit
+def compute_mu(trend, X, betas, s_m, s_a):
+    return jnp.multiply(trend, (1 + X.dot(betas * s_m))) + X.dot(betas * s_a)
+
+
+def model_linear(s_m, s_a, sigmas, X, cap, t, t_change, tau, T, S, y, changepoint):
+    k = sample("k", dist.Normal(0, 5))
+    m = sample("m", dist.Normal(0, 5))
+    with plate('delta_dim', S):
+        delta = sample("delta", dist.Laplace(0, tau))
+        trend = deterministic("trend", linear_trend(k, m, delta, t, changepoint, t_change))
+    sigma_obs = sample("sigma_obs", dist.HalfNormal(scale=0.5))
+    betas = sample("beta", dist.MultivariateNormal(0, jnp.diag(sigmas)))
+    mu = deterministic("mu", compute_mu(trend, X, betas, s_m, s_a))
+    with plate("data", T):
+        sample("obs", dist.Normal(mu, sigma_obs), obs=y)
+
+
+def model_logistic(s_m, s_a, sigmas, X, cap, t, t_change, tau, T, S, y, changepoint):
```

Comment:
Great question! I originally wanted to do something like what you're suggesting but the way the jax compilation works, it would actually evaluate both sides of the if statement:

```
the lax.cond form always computes both sides of the if statement before choosing which one to carry forward, while the true if statement changes its computation based on the conditional.
```

Source: https://www.stochasticlifestyle.com/useful-algorithms-that-are-not-optimized-by-jax-pytorch-or-tensorflow/